### DISCUSSION TITLE
Add kubernetes cluster using service account token with read-only permissions

### DISCUSSION THREAD

**Discussion by alexevansvu** [mia-platform/community#379] [Q&A]
### Which section?

Console

### Which is the version?

1.91.6

### Description

We're trying to add a Kubernetes cluster connection to a company so as to populate runtime data, but we want it to have read-only permissions. 

I've confirmed that our configuration (cluster URL, certificate authority, and access token) are all valid, and the token works for read operations via Kubernetes kubectl. 

However, when I try to add the cluster in mia-console, I get the following error:
![image](https://github.com/mia-platform/community/assets/127889749/557892d4-6e0b-4717-935a-84ae7b0ee980)

My questions are: 

1. Is adding a read only cluster connection supported?
2. If so, are there a specific set of permissions Mia is looking for on the access token?

**Comment by davidebianchi** [comment]
Hi @alexevansvu!
Yes, it is supported to add readonly connection. To connect the cluster, the service make 2 different API call: one to know the version of the cluster and the second to take the nodes information.
This second call should be optional, but from the error message it seems to be the error message of this second one.
Does your access token have permissions to take information about the nodes?

Can you tell me what is the response code and message of the create cluster api?

**Comment by alexevansvu** [comment]
Hi @davidebianchi,

Thank you for the quick response! 

The token does not have permissions to read information about the nodes, and I am currently working to get the response code and message from the API

**Comment by alexevansvu** [comment]
Small update: adding read access to the nodes to the service account access token took care of getting the cluster to connect successfully, however it does not appear to be pulling information on running pods in the runtime page for a given project under the company the connection was added to.

**Reply by davidebianchi** [reply]
Oh, thank you for the update! We will investigate since it should work also without node permission!
What permissions does the role have? There are resources on the specific namespace linked to the project environment?

**Reply by alexevansvu** [reply]
Project Environment configuration turned out to be the issue. 

Everything was running in the right place on the kubernetes side, but we needed to go in to CMS and update the environment configuration JSON for the project so that it uses the new cluster connection's Cluster ID. 

Everything is working as expected now

Thank you for your help!

**Reply by Danielecina** [reply]
hello, @alexevansvu I'm glad the problem is back, I would like to ask you if you could take a few more minutes to describe the steps you took to configure the token up to the creation on the project of the environment 
 
Could you please let me know why you changed the clusterId of the project in CMS? For convenience?
Wasn't it sufficient to change the cluster token on the cluster page? Did you create a new cluster and want to assign it to the already configured environment?

**Reply by Beridar** [reply]
I'm on @alexevansvu 's team; I'll speak to part of that.

When we first configured our new Mia install, we had some internal concerns about cluster permissions we really needed to give to the Mia Platform install.  It's been our plan all along to use the [Enhanced Project Workflow](https://docs.mia-platform.eu/docs/development_suite/set-up-infrastructure/enhanced-project-workflow) for deployment.  With that model, we couldn't say why our Mia Platform install needed an active connection to the cluster -- and so we set up a connection to a sandbox cluster just to keep moving forward.

It was not ideal, but it got us to move forward and we are able to deploy without issue.  :tada:

Making the change to a read-only token is a step in the right direction for us.  Because we were using the enhanced project workflow to handle deployments, we were deploying to one cluster but had the environments on a project configured to read from another cluster.  We had a different cluster (*with a different clusterId, token, and URL*), and **that** prompted changing things in the CMS.
